copyoptions.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oras
+
+import (
+	"context"
+	"math"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/semaphore"
+)
+
+const defaultConcurrencyLimit = 3
+
+type copyOptions struct {
+	limiter             *semaphore.Weighted
+	copyCallbackHandler func(ctx context.Context, desc ocispec.Descriptor) error
+}
+
+func NewCopyOptions() *copyOptions {
+	return &copyOptions{
+		limiter: semaphore.NewWeighted(defaultConcurrencyLimit),
+	}
+}
+
+func (co *copyOptions) WithConcurrencyLimit(concurrency int64) *copyOptions {
+	co.limiter = semaphore.NewWeighted(concurrency)
+	return co
+}
+
+func (co *copyOptions) WithCopyCallbackHandler(fn func(ctx context.Context, desc ocispec.Descriptor) error) *copyOptions {
+	co.copyCallbackHandler = fn
+	return co
+}
+
+type extendedCopyOptions struct {
+	*copyOptions
+	depth               uint32
+	upEdgeFilterHandler func(ctx context.Context, desc ocispec.Descriptor) (bool, error)
+	customUpEdgesFinder func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error)
+}
+
+func NewExtendedCopyOptions() *extendedCopyOptions {
+	return &extendedCopyOptions{
+		copyOptions: NewCopyOptions(),
+		depth:       math.MaxUint32,
+	}
+}
+
+func (eco *extendedCopyOptions) WithConcurrencyLimit(concurrency int64) *extendedCopyOptions {
+	eco.limiter = semaphore.NewWeighted(concurrency)
```

Comment:
how about `eco.copyOptions.WithConcurrencyLimit(concurrency)`

Reply:
`copyOptions` is embeded in `extendedCopyOptions`. So `eco.copyOptions.WithConcurrencyLimit(concurrency)` is equivalent to `eco.WithConcurrencyLimit(concurrency)`.

Reply:
I think @qweeah is referring

```go
func (eco *extendedCopyOptions) WithConcurrencyLimit(concurrency int64) *extendedCopyOptions {
    eco.copyOptions.WithConcurrencyLimit(concurrency)
    return eco
}
```

Reply:
I also noticed the drawback of this pattern. `eco` is not able to reuse the `WithConcurrencyLimit` function of `copyOptions`, and thus we have redundant codes. 

Thus, I suggest dropping the fluent-style interface. Instead, we can have

```go
func (co *copyOptions) LimitConcurrency(concurrency int64)
```

Another drawback for the fluent-style interface is the error handling. What if you have

```go
func (eco *extendedCopyOptions) WithMagic(magic any) (*extendedCopyOptions, error)
```

Then we are not able to chain the options.